### PR TITLE
[Flexible Header] Rework our enforcement of content insets.

### PR DESCRIPTION
**Pull request by jverkoey**
Essential changes in this PR:

- Rename `fhv_addInsetsToScrollView` to `fhv_enforceInsetsForScrollView`. The implementation can now be called repeatedly to ensure that the scroll view's content insets match our desired insets.
- Generally trying to remove conditional bail-outs and relying more on code paths that can be executed in any logical condition. E.g. avoiding `if (_isChangingStatusBarVisibility) { return; }`.
- Removed `fhv_adjustInsetsForSafeAreaInScrollView` in favor of `fhv_enforceInsetsForScrollView`.
- Remove `_currentSafeAreaInsetTop` state and checks in favor of always enforcing the content insets.

### CODE REVIEW

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m
Comment (chuga) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: But we don't want to do any of this when min or max height has been explicitly set and _minMaxHeightIncludesSafeArea hasn't been set to NO :(

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m
Comment (chuga) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: Try this with the App Bar example that uses a TableViewController.

**Comment by jverkoey** [comment]
This needs to be worked on a bit more and tested rigorously. The essential idea is to ensure that `contentInset.top` always equals maximumHeight + any client-provided inset, so that even on non-X devices when the status bar is hidden and the content inset adjustment behavior is enabled, our contentInset doesn't shrink in reaction to UIKit changing the adjustedContentInset.

**Comment by randallli** [comment]
Looking out for changes in behaviour...
It seems that when you hide the status bar it still considers the 20 pixels as part of the min height. This didn't use to be the case.

before
simulator screen shot - iphone 8 - 2017-10-24 at 11 15 54

after
simulator screen shot - iphone 8 - 2017-10-24 at 11 16 28

**Comment by randallli** [comment]
featherless@ walked me through it and the screenshot differences that i just pointed out above is the desired change.

**Comment by jverkoey** [comment]
@chuga https://github.com/material-components/material-components-ios/pull/2264 has a description.

**Comment by jverkoey** [comment]
Blocked by #2264 and #2265.

**Comment by jverkoey** [comment]
This is ready for review.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m
Comment (chuga) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m:602]: As discussed over chat, I think we still need this. This helps us catch any changes to the safe area inset that we might've missed for things like [UIApplication mdc_safeSharedApplication].keyWindow not being set when safeAreaInsetsDidChange gets called.

Comment (jverkoey) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView.m:602]: Added back.

**Review by chuga — APPROVED** [review]
I see no regressions

**Comment by jverkoey** [comment]
Will merge once I get @randallli approval.

**Review by randallli — APPROVED** [review]

**Pull request merged**
